packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,225 @@
+import { Tag } from '../cloudformation/tag';
+import { Construct } from './construct';
+import { Token } from './tokens';
+
+/**
+ * ITaggable indicates a entity manages tags via the `tags` property
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * Properties Tags is a dictionary of tags as strings
+ */
+export interface Tags {
+    [key: string]: string
+}
+
+/**
+ * A an object of tags with value and properties
+ *
+ * This is used internally but not exported
+ */
+interface FullTags {
+    [key: string]: {value: string, props?: TagProps};
+}
+
+/**
+ * Properties for a tag
+ */
+export interface TagProps {
+    /**
+     * If true all child taggable `Constructs` will receive this tag
+     *
+     * @default true
+     */
+    propagate?: boolean;
+
+    /**
+     * If set propagated tags from parents will overwrite tags in children
+     *
+     * @default false
+     */
+    propagateOverride?: boolean;
+
+    /**
+     * If set this tag will overwrite existing tags
+     *
+     * @default true
+     */
+    overwrite?: boolean;
+}
+
+/**
+ * Properties for removing tags
+ */
+export interface RemoveProps {
+    /**
+     * If true prevent this tag form being set via propagation
+     *
+     * @default true
+     */
+    blockPropagate?: boolean;
+}
+
+/**
+ * TagManager facilitates a common implementation of tagging for Constructs.
+ *
+ * Each construct that wants to support tags should implement the `ITaggable`
+ * interface and properly pass tags to the `Resources` (Cloudformation) elements
+ * the `Construct` creates using `toCloudformation()` for lazy evaluations
+ */
+export class TagManager {
+
+    public static readonly DEFAULT_TAG_PROPS: TagProps = {
+        propagate: true,
+        propagateOverride: false,
+        overwrite: true
+    };
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    /*
+     * Internally tags will have properties set
+     */
+    private readonly _tags: FullTags = {};
+
+    /*
+     * Tags that will be reomved during `tags` method
+     */
+    private readonly blockedTags: string[] = [];
+
+    constructor(private readonly parent: Construct, initialTags: Tags  = {}) {
+        for (const key of Object.keys(initialTags)) {
+            const tag = {value: initialTags[key],
+                props: {propagate: true, propagateOverride: false}};
+            this._tags[key] = tag;
+        }
+    }
+
+    /**
+     * Converts the `tags` to a Token for use in lazy evaluation
+     */
+    public toCloudformation(): Token {
```

Comment:
Typo `toCloudFormation()`, but again I would find this name slightly misleading since it's not actually a `StackElement` that's participating in the `toCloudFormation()` rendering as usual, and in fact the returned object does not have the same shape.

Maybe call it something like `renderTags()` or `asTagObject()`, `toTagJson()` or similar instead.

Reply:
You might also make `TagManager extend Token` and implement `resolve()` to return `this.toArray()`. Then you can use the object everywhere you would pass `tags` directly to CloudFormation objects.

There is precedent for this: we do the same for `PolicyDocument`, for example.

Reply:
Agree on the typo (🤦‍♂️ ). I think I like `renderTags` better. Let me look at policy document because I was considering the `extends Token`. Is #518 going to change anything here?

Reply:
No, #518 is only about embedding `Token`s representing string values into actual `string` types (which becomes possible once you `extends Token`, but we're not going to use that in any case).

Reply:
I think I want to take a stab at `extends Token` I'll separate this into a separate commit after fixing some of the naming that we can revert if we have disagreements.